### PR TITLE
feat: refactor/improve output capturing mechanism

### PR DESCRIPTION
**Pull request by taylordotfish**
This PR updates Atuin's command output capturing code. Much of this is to lay the groundwork for future changes, but there are also some immediate benefits:

* OSC 133 parsing has been improved. The parser is now **fully allocation-free**, and bugs relating to escape sequences that spanned chunk boundaries have been fixed.
* Output captures **match reality much more closely**. Previously, we did not correctly handle output that included certain kinds of escape sequences, like those that move the cursor to an absolute position. Now we use a more robust capture mechanism.
* SGR escape sequences (responsible for colors and formatting) are now included in captured output. **This enables Atuin AI to answer questions like “what does the red text mean in the command I just ran?”.**

Atuin AI can now answer questions like this:

atuin-ai-colors

### CODE REVIEW

File: crates/atuin-pty-proxy/src/runtime.rs [hunk 1]
```diff
@@ -72,53 +70,44 @@ fn run(options: RuntimeOptions) -> eyre::Result<()> {
...
-                        (capture_tracker.as_mut(), options.command_capture_sink.as_ref())
-                    {
-                        tracker.push(&buf[..n], sink);
-                    }
-
-                    if let Some(highlighter) = highlighter.as_mut() {
-                        let rendered = highlighter.render(&buf[..n]);
-                        let _ = msg_tx.send(Msg::Data(rendered.clone()));
+                    let raw_data = &buf[..n];
+                    let _ = msg_tx.send(Msg::Data(raw_data.to_vec()));
```
[the remark this answers is not in the document]
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/runtime.rs:99]: Yes, the capture and snapshot emulators run serially, which could under high load block the stdout thread, but at the same time, the status quo before this PR is that the capture emulator runs directly on the stdout thread itself, which means it *always* blocks the stdout thread, which is arguably worse. Under this PR, bursts can be absorbed without delaying the stdout thread at all; before this PR, *any* burst would unconditionally block stdout forwarding. Running the emulators on a separate thread is the cleaner design; if we find that running the two emulators serially is too slow and causes the stdout thread to block under reasonable workloads, we can explore increasing the MPSC queue buffer size and/or running the emulators on separate threads. We should not make those decisions preeemptively, as they will increase memory, and we need to be conscious of memory use in the PTY proxy as well.

File: crates/atuin-common/src/string/trim.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
+mod sealed {
+    pub trait Sealed {}
+}
+
+/// A pattern that, unlike `std::str::pattern::Pattern`, does not consume the pattern when used.
```
Comment (markovejnovic) [inline crates/atuin-common/src/string/trim.rs]:
```suggestion
/// A pattern that, unlike [`std::str::pattern::Pattern`], does not consume the pattern when used.
```

File: crates/atuin-common/src/string/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
+use std::fmt;
+
+/// A string buffer limited to a certain length.
+#[derive(Clone)]
+pub struct BoundedBuffer {
```
Comment (markovejnovic) [inline crates/atuin-common/src/string/buffer.rs:5]: Seems difficult to generalize this. I tried, but ultimately failed. Not sure it's applicable here, but are you interested in making this accept `atuin_common::string::Measure`?

File: crates/atuin-common/src/string/buffer.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+        self.inner.push_str(&s[..s.floor_char_boundary(available)]);
+        self.truncated = true;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (markovejnovic) [inline crates/atuin-common/src/string/buffer.rs:90]: Some vibe-coded proptests could be good here to catch edge cases

File: crates/atuin-common/src/string/trim.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            }
+
+            fn trim_end_matches<'a>(&mut self, s: &'a str) -> &'a str {
+                s.trim_end_matches(($to_pattern)(self))
+            }
+        }
+    };
+}
+
+// Implement `PatternRef` for all of the types that implement `Pattern`.
```
Comment (markovejnovic) [inline crates/atuin-common/src/string/trim.rs:32]:
```suggestion
```

I'm guessing blanket impls didn't work?

File: crates/atuin-daemon/proto/semantic.proto [hunk 6]
```diff
@@ -7,8 +7,13 @@ service Semantic {
 }
 
 message CommandCapture {
+  // The shell prompt, as rendered by the terminal. Contains SGR escape
+  // sequences.
   string prompt = 1;
+  // The command typed by the user, as rendered by the terminal. Contains SGR
+  // escape sequences.
```
Comment (markovejnovic) [inline crates/atuin-daemon/proto/semantic.proto]:
```suggestion
  // The command typed by the user, as rendered by the terminal. May contain SGR
  // escape sequences.
```

File: crates/atuin-daemon/proto/semantic.proto [hunk 7]
```diff
@@ -7,8 +7,13 @@ service Semantic {
 }
 
 message CommandCapture {
+  // The shell prompt, as rendered by the terminal. Contains SGR escape
+  // sequences.
```
Comment (markovejnovic) [inline crates/atuin-daemon/proto/semantic.proto]:
```suggestion
  // The shell prompt, as rendered by the terminal. May contain SGR escape
  // sequences.
```

File: crates/atuin-pty-proxy/src/screen.rs [hunk 8]
```diff
@@ -22,23 +25,65 @@ pub fn socket_path() -> Result<PathBuf, SecureTempDirError> {
...
+pub fn spawn_parser_thread(rows: u16, cols: u16, screen_rx: Receiver<Msg>, options: ParserOptions) {
+    std::thread::spawn(move || {
+        let rows = NonZeroU16::new(rows).unwrap_or(NonZeroU16::MIN);
+        let cols = NonZeroU16::new(cols).unwrap_or(NonZeroU16::MIN);
+        let mut parser = Parser::new(rows, cols, options);
+        for msg in screen_rx {
+            parser.handle_msg(msg);
+        }
     });
 }
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/screen.rs]: I generally prefer renaming `Parser` to be `ParserWorker`, while making a public-facing `Parser`.

```rust
struct Parser {
  thread: JoinHandle<()>,
}

impl Parser {
  pub fn spawn((rows, cols), screen_rx, options) -> Self;
}
```

That way the thread isn't detached -- that is its own can of worms. If you want, you can do this in this PR, or I can do it in a follow-up when I sweep through all of the pty proxy.

File: crates/atuin-pty-proxy/src/screen.rs [hunk 9]
```diff
@@ -22,23 +25,65 @@ pub fn socket_path() -> Result<PathBuf, SecureTempDirError> {
...
+pub struct ParserOptions {
+    pub sink: Option<CommandCaptureSink>,
+    pub debug_osc133: bool,
+}
 
-        loop {
-            let Ok(first) = msg_rx.recv() else {
-                break;
-            };
+struct Parser {
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/screen.rs:34]: This name is very generic. While you're at it, I would strongly suggest you rename this to something like `ScreenParser` (even that kind of sucks). Would suggest coming up with a good name.

File: crates/atuin-pty-proxy/src/screen.rs [hunk 10]
```diff
@@ -22,23 +25,65 @@ pub fn socket_path() -> Result<PathBuf, SecureTempDirError> {
...
+    emulator: vt100::Parser,
+    tracker: Option<CommandCaptureTracker>,
+    highlighter: Option<Osc133DebugHighlighter>,
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/screen.rs:37]:
```suggestion
    osc133_highlighter: Option<Osc133DebugHighlighter>,
```

File: crates/atuin-pty-proxy/src/screen.rs [hunk 11]
```diff
@@ -22,23 +25,65 @@ pub fn socket_path() -> Result<PathBuf, SecureTempDirError> {
...
+                    highlighted = highlighter.render(&raw_data);
+                    &highlighted
+                } else {
+                    &raw_data
+                };
+                self.emulator.process(data);
+            }
+            Msg::Resize { rows, cols } => {
+                let rows = NonZeroU16::new(rows).unwrap_or(NonZeroU16::MIN);
+                let cols = NonZeroU16::new(cols).unwrap_or(NonZeroU16::MIN);
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/screen.rs:70]: Hmmm. Is it possible to have a terminal with zero columns and zero rows? `crossterm::terminal::size()` returns `(u16, u16)`.

- If that is not possible, I'd recommend moving this logic into the message (so the message carries `NonZeroU16`. You can `debug_assert` that this invariant is held in `atuin-pty-proxy/src/runtime.rs:160`.
- If that is possible, then you can leave the things here as-is, though I would make a note mentioning _why_ the Vt100 parser cannot accept col=0, row=0.

File: crates/atuin-pty-proxy/src/runtime.rs [hunk 12]
```diff
@@ -160,7 +149,6 @@ fn run(options: RuntimeOptions) -> eyre::Result<()> {
 fn spawn_resize_handler(
     master: Box<dyn portable_pty::MasterPty + Send>,
     resize_tx: mpsc::SyncSender<Msg>,
-    current_cols: Arc<AtomicU16>,
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/runtime.rs:163]: 👍

File: crates/atuin-pty-proxy/src/runtime.rs [hunk 13]
```diff
@@ -72,53 +70,44 @@ fn run(options: RuntimeOptions) -> eyre::Result<()> {
...
                     } else {
-                        let _ = msg_tx.send(Msg::Data(buf[..n].to_vec()));
+                        raw_data
+                    };
 
-                        if stdout.write_all(&buf[..n]).is_err() {
-                            break;
-                        }
+                    if stdout.write_all(data).is_err() {
+                        break;
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/runtime.rs:110]: I know this was here before you, but this `break` terminates the thread. Does this not leave pty-proxy in an inconsistent state, wherein the debug highlighter just stops working?

Won't block your PR on this but interested in your thoughts.

File: crates/atuin-pty-proxy/src/debug.rs [hunk 14]
```diff
@@ -15,28 +15,22 @@ impl Osc133DebugHighlighter {
...
-            rendered.extend_from_slice(event_label(located.event));
+        for chunk in chunks {
+            rendered.extend_from_slice(chunk.data);
+            rendered.extend_from_slice(event_label(chunk.event));
             rendered.extend_from_slice(RESET);
-            start = offset;
         }
 
-        rendered.extend_from_slice(&data[start..]);
+        rendered.extend_from_slice(chunk_iter.trailing_data());
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/debug.rs:33]: 👍

File: crates/atuin-pty-proxy/src/osc133.rs [hunk 15]
```diff
@@ -168,212 +292,140 @@ impl Default for Parser {
 
 impl Parser {
```
Comment (markovejnovic) [inline crates/atuin-pty-proxy/src/osc133.rs:289]: Have you considered using [`winnow`](https://docs.rs/winnow/latest/winnow/) for this? This code reads very much like a hand-rolled combinatorial parser and I can't help but wonder if using a combinatorial parser library would have been helpful here.

**Review by markovejnovic — APPROVED** [review]
Mostly nits. I haven't had a chance to review `osc133.rs` -- it's a parser and that's not easy to review.

Fortunately, I don't really care for it -- it's a debugging utility users don't even hit. Arguably, I think we should tear it out behind a `debug`-only build.

File: crates/atuin-pty-proxy/src/runtime.rs (hunk 12 above)
Comment (markovejnovic) [reply crates/atuin-pty-proxy/src/runtime.rs:163]: Also I feel the ordering may have been wrong here -- probably needed `acq/rel`, so I think this change is good.

File: crates/atuin-common/src/string/trim.rs (hunk 2 above)
Comment (taylordotfish) [reply crates/atuin-common/src/string/trim.rs]: I'll test if this works. I was unsure since it's an unstable item, but maybe you can link to it still.

File: crates/atuin-common/src/string/buffer.rs (hunk 3 above)
Comment (taylordotfish) [reply crates/atuin-common/src/string/buffer.rs:5]: Yes, I think that could be a good idea. Probably will push that to a followup PR if that's ok.

File: crates/atuin-common/src/string/trim.rs (hunk 2 above)
Comment (taylordotfish) [reply crates/atuin-common/src/string/trim.rs]: Yep, works

File: crates/atuin-common/src/string/trim.rs (hunk 5 above)
Comment (taylordotfish) [reply crates/atuin-common/src/string/trim.rs:32]: Yes. We can't name `std::str::pattern::Pattern` in our code because it's unstable.

File: crates/atuin-pty-proxy/src/screen.rs (hunk 9 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs:34]: I think it's a matter of style. The fact that `Parser` is in the `screen` module clarifies it enough for me (`screen::Parser`), but I don't feel that strongly. If you do feel strongly about it I'm fine making the change.

File: crates/atuin-pty-proxy/src/screen.rs (hunk 10 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs:37]: I mildly disagree with this; I think the type makes it clear, and it was already called `highlighter` in our existing code in runtime.rs. But if you feel strongly about it I don't mind changing it.

File: crates/atuin-pty-proxy/src/screen.rs (hunk 11 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs:70]: I think it is possible. I'm happy to add a comment; however `atuin_vt100::Parser::new` takes `NonZeroU16`s, so the fact that we can't pass 0 is already enforced by the type system. If we were using upstream `vt100` I'd agree with you that we need a comment to explain that we can't pass 0, but now the compiler enforces it.

That said, I don't have an objection to adding a comment explicitly stating this.

File: crates/atuin-pty-proxy/src/runtime.rs (hunk 12 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/runtime.rs:163]: Yes, I believe there was a race condition. I'm not even sure `acq/rel` would have fixed it honestly.

File: crates/atuin-pty-proxy/src/screen.rs (hunk 8 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs]: I'll make a similar change. I think `Parser` is a fine name because it *is* the actual parser. In my mind, the parser thread owns the parser. I also prefer having the parser thread be named something that communicates it's a thread.

I'll make a `ParserThread` type that owns the `JoinHandle`. I'll also have it be responsible for creating and holding the `SyncSender`; there's no reason why the caller should have to do that manually, and then `ParserThread` can have a simple method called `send` that sends a message.

If you're not satisfied with that design, we can talk about it and address in a followup PR like you said.

Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs]: Never mind, that won't work because we need to clone the `SyncSender` and use it in multiple places.

Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs]: I'm going to leave this change for later so we can discuss, but for now, I'll make `spawn_parser_thread` return a `JoinHandle`. I won't do anything with it in this PR but I'll bind it to an underscore-prefixed variable name in `runtime.rs`.

File: crates/atuin-pty-proxy/src/runtime.rs (hunk 13 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/runtime.rs:110]: If we fail writing to the parent terminal, it's almost certainly because the FD got closed, which means *everything* is broken. Not just the debug highlighter—we can't forward data from the PTY either…

File: crates/atuin-pty-proxy/src/osc133.rs (hunk 15 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/osc133.rs:289]: I haven't. I'm open to it in theory but have questions:
* How does the performance of `winnow` compare? Is it allocation-free?
* The grammar of our parser is technically `*`. Nothing is actually a parse error. Is `winnow` a good choice for this?
* How does `winnow` handle streaming? We push arbitrary-sized chunks to our parser, and OSC 133 escape sequences can cross chunk boundaries.
* Our parser is quite simple, really. OSC 133 sequences aren't hard to parse. Is the added complexity of `winnow` worth it?

**Comment by taylordotfish** [comment]
> Mostly nits. I haven't had a chance to review `osc133.rs` -- it's a parser and that's not easy to review.
> 
> Fortunately, I don't really care for it -- it's a debugging utility users don't even hit. Arguably, I think we should tear it out behind a `debug`-only build.

@markovejnovic: To clarify, `osc133.rs` actually isn't debug-only. It's used by `capture.rs` to figure out where command output starts and ends.

File: crates/atuin-pty-proxy/src/screen.rs (hunk 9 above)
Comment (taylordotfish) [reply crates/atuin-pty-proxy/src/screen.rs:34]: The more I think about it, I feel a little more strongly that we shouldn't make this change. It's a private type, and it's in `screen.rs` so I think it's clear it's a “screen parser”. I generally dislike redundancy in names, and it's a pattern that's discouraged in Rust.
